### ISSUE TITLE
Build agentic source alignment skill for Mondo — ICD11 Foundation as first target

### ISSUE THREAD

**Issue by matentzn** [monarch-initiative/mondo-ingest#894]
Develop a **Claude Code skill** for LLM-assisted ontology alignment that can propose high-confidence Mondo-to-X mappings. The first target is ICD11 Foundation. The bar is >95% precision on proposed mappings — clinicians will review the output.

Workflow wise I want to be able to start a new session, and just say 

```
/mondo-align icd11 foundation
```

In the context of this repo, and get as an output a SSSOM file with new mappings to add and rich metadata (justifications).

## Context

Coverage for ICD11 and ICD10 mappings is incomplete in Mondo and expanding it manually is slow. Issue #9684 demonstrated that simple transitive inference via semra can produce ~975 new ICD11/ICD10 mappings, but that approach is limited to cases where mapping chains already exist.

LLM-based ontology alignment is an active research area but agents have historically been weak at this — they hallucinate mappings, miss granularity differences, and struggle with the 1:many / many:1 cases that dominate disease ontology alignment. A skill that encodes the [Mondo Mapping Curation Guide](https://docs.google.com/document/d/1irMiD5Syf0CPlfFjEj__5kgb0RJZOGpDw7C7v6Ic0Gs/edit?tab=t.0) as structured agent instructions could change this.

### Relevant literature (2024–2026)

Recent work shows LLM-based alignment can match or exceed traditional systems when done right:

| System | Key idea | Best result |
|---|---|---|
| **[MILA](https://arxiv.org/abs/2501.11441)** (2025, Information Fusion) | RAG + prioritized DFS; LLM only for borderline cases (90% fewer LLM calls) | F1 0.83–0.95 on OAEI biomedical; SOTA on 4/5 unsupervised tasks |
| **[Agent-OM](https://arxiv.org/abs/2312.00326)** (VLDB 2024) | Two Siamese LLM agents + CoT with self-consistency | Best P/R/F1 on OAEI Anatomy non-trivial correspondences |
| **[LLMs4OM](https://arxiv.org/abs/2404.10317)** (ESWC 2024) | RAG candidate retrieval + zero-shot matching; concept-children context | 94.3% F1 on MI-EMMO |
| **[LLMs as Oracles](https://arxiv.org/abs/2508.08500)** (EACL 2026) | LLM validates only uncertain correspondences from traditional OM | Top-2 on OAEI 2025 Bio-ML track |
| **[GPT-4o medical mapping](https://www.frontiersin.org/journals/artificial-intelligence/articles/10.3389/frai.2025.1546179/full)** (Frontiers 2025) | BioBERT embeddings + ChromaDB RAG | 93.75% precision, 96.26% F1 |
| **[GPT-4 + RAG for SNOMED CT](https://pubmed.ncbi.nlm.nih.gov/40380580/)** (2025) | RAG beats fine-tuning for clinical terminology mapping | 96.2% valid match rate |

Key patterns from the literature:
1. **RAG is non-negotiable** — every top system retrieves ontology context before prompting
2. **Selective LLM invocation** — use string/embedding similarity for easy matches, reserve LLM for ambiguous cases (MILA cut calls by 90%)
3. **Hierarchical context** — providing parent/children terms in prompts dramatically improves disambiguation
4. **CoT + self-consistency validation** — reduces hallucinated mappings (Agent-OM)
5. **Known failure modes** — the [OAEI-LLM benchmark](https://arxiv.org/abs/2409.14038) catalogs LLM hallucination types: false mappings, rank confusion (align-up/align-down), disputed mappings

## Expected behaviour

A skill that, given a target source (e.g. ICD11 Foundation), produces a set of candidate Mondo mappings in SSSOM format with:

- **>95% precision** on mappings the agent deems high-confidence
- Clear confidence tiers (auto-accept, needs-review, rejected) so clinician review is efficient
- Full provenance: which evidence chains, metadata lookups, and disambiguation steps led to each mapping
- Output as a reviewable SSSOM TSV that can be loaded into Mondo's mapping pipeline

### Phase 1: Mapping chain bootstrapping

1. Collect existing Mondo mappings from [Mondo ingest](https://github.com/monarch-initiative/mondo-ingest/blob/main/src/mappings) and the normal mondo.sssom.tsv (SSSOM files)
2. Collect available external mapping sets (WHO ICD11-to-ICD10 mappings, UMLS crosswalks, etc.) - see Mondo
3. Chain mappings transitively to generate candidate Mondo-ICD11 Foundation pairs (using semra, as in #9684)
4. Score candidates by chain length, mapping predicate consistency, and source reliability. Start with chain length 1 (1 hop)
5. You can also pull in lexical matches for every source, eg (please **do not rebuild a lexical pipeline** - we have that):
     a. https://github.com/monarch-initiative/mondo-ingest/blob/main/src/ontology/lexmatch/unmapped_doid_lex.tsv
     b. https://github.com/monarch-initiative/mondo-ingest/blob/main/src/ontology/lexmatch/unmapped_doid_lex_exact.tsv

### Phase 2: Metadata enrichment and disambiguation

Warning: keep it agentic: let Claude figure out what to do rather than you being overly specific about it!

For each candidate mapping pair:

1. Retrieve Mondo term metadata: definition, synonyms, parent classes, logical axioms, existing xrefs
2. Retrieve ICD11 Foundation metadata: title, definition, parent codes, inclusions/exclusions, index terms
3. Look up disease information from trusted sources (OMIM, PubMed, Orphanet) when needed for disambiguation
4. Compare granularity — detect 1:many or many:1 situations where Mondo and ICD11 split/lump differently

### Phase 3: LLM-based alignment decision

Implement the Mondo Mapping Curation Guide as structured agent logic:

1. **Exact match detection**: Same disease concept at same granularity → `skos:exactMatch`
2. **Broad/narrow detection**: One concept subsumes the other → `skos:broadMatch` / `skos:narrowMatch`
3. **Related match**: Overlapping but not subsumptive → `skos:relatedMatch`
4. **No match**: Confidently unrelated → skip
5. **Uncertain**: Not enough evidence → flag for human review

Use chain-of-thought reasoning with explicit evidence citation. The agent should explain *why* it chose a particular mapping predicate, not just output a label.

### Phase 4: Confidence calibration

- Run the skill against Mondo terms that *already have* ICD11 mappings (held-out validation set)
- Measure precision/recall against the gold standard
- Tune confidence thresholds until the "auto-accept" tier hits >95% precision
- Document false positive patterns so the skill can be iteratively improved

### Key design principles

- **Err toward rejection**: A missed mapping is cheap; a wrong mapping wastes clinician time and erodes trust
- **Provenance is non-negotiable**: Every proposed mapping must trace back to specific evidence (eg. a MappingChain, see SSSOM metadadata standard)
- **Follow the literature**: MILA's selective-invocation pattern and Agent-OM's CoT-SC validation are directly applicable
- **Iterative**: Start with the easy cases (identical labels, clear 1:1 matches) and progressively tackle harder ones

## Open questions

- What external ICD11 metadata sources are accessible programmatically? (WHO API, UMLS, etc.)
- Should the skill chain through ICD10 → ICD11 mappings from WHO as a starting point, or start fresh?
- How to handle ICD11 Foundation vs. ICD11 MMS (linearization) — Foundation is the target but MMS codes are what clinicians use

## Related

- #9684 — Add 975 new ICD11/ICD10 mappings through simple semra inference (prior art on transitive chaining)
- #5410 — Mondo mappings internal representation
- #10130 — Separate hop provenance from semantic precision on xref source annotations
- [Mondo Mapping Curation Guide](https://docs.google.com/document/d/1irMiD5Syf0CPlfFjEj__5kgb0RJZOGpDw7C7v6Ic0Gs/edit?tab=t.0)
- [Agent-OM GitHub](https://github.com/qzc438/ontology-llm) — reference implementation for agentic ontology matching
- [LLMs4OM GitHub](https://github.com/HamedBabaei/LLMs4OM) — RAG-based ontology matching framework
- [OAEI-LLM hallucination benchmark](https://arxiv.org/abs/2409.14038) — taxonomy of LLM failure modes in OM
- [LLMs in Bio-Ontology Research survey](https://pmc.ncbi.nlm.nih.gov/articles/PMC12649945/) — comprehensive review

**Comment by cmungall** [comment]
See also https://github.com/monarch-initiative/boomer-py

**Comment by matentzn** [comment]
@cmungall we are exploring this separately in #885 with @ptgolden.

I think both pipelines can be developed separately - I will marry them when I see the working boomer pipeline right there in front of me and a draft of a `/mondo-align` skill. I dont want to block @tannerzhang here because there is a huge learning curve on boomer-py.
